fury/animation.py

```diff
@@ -0,0 +1,744 @@
+            k1["cp"] = self.keyframes['post_cp'][t_s]
+            k2["cp"] = self.keyframes['pre_cp'][t_e]
+        return {"start": k1, "end": k2}
+
+    @staticmethod
+    def _lerp(v1, v2, t1, t2, t):
+        if t1 == t2:
+            return v1
+        v = v2 - v1
+        dt = (t - t1) / (t2 - t1)
+        return dt * v + v1
+
+    @staticmethod
+    def _get_time_delta(t, t1, t2):
+        return 0 if t <= t1 else 1 if t >= t2 else (t - t1) / (t2 - t1)
+
+
+class StepInterpolator(Interpolator):
+    """Step interpolator for keyframes.
+
+    This is a simple step interpolator to be used for any shape of
+    keyframes data.
+    """
+
+    def __init__(self, keyframes):
+        super(StepInterpolator, self).__init__(keyframes)
+        self.id = 0
+
+    def setup(self):
+        super(StepInterpolator, self).setup()
+
+    def interpolate(self, t):
+        t_lower = self._get_nearest_smaller_timestamp(t, include_last=True)
+        return self.keyframes[t_lower]['value']
+
+
+class LinearInterpolator(Interpolator):
+    """Linear interpolator for keyframes.
+
+    This is a general linear interpolator to be used for any shape of
+    keyframes data.
+    """
+
+    def __init__(self, keyframes=None):
+        if keyframes is None:
+            keyframes = {}
+        super(LinearInterpolator, self).__init__(keyframes)
+        self.id = 1
+
+    def setup(self):
+        super(LinearInterpolator, self).setup()
+
+    def interpolate(self, t):
+        t1 = self._get_nearest_smaller_timestamp(t)
+        t2 = self._get_nearest_larger_timestamp(t)
+        p1 = self.keyframes[t1]['value']
+        p2 = self.keyframes[t2]['value']
+        return self._lerp(p1, p2, t1, t2, t)
+
+
+class SplineInterpolator(Interpolator):
+    """N-th degree spline interpolator for keyframes.
+
+    This is a general n-th degree spline interpolator to be used for any shape
+    of keyframes data.
+    """
+
+    def __init__(self, keyframes, degree=3, smoothness=3):
+        self.degree = degree
+        self.smoothness = smoothness
+        self.tck = []
+        self.linear_lengths = []
+        super(SplineInterpolator, self).__init__(keyframes)
+        self.id = 6
+
+    def setup(self):
+        super(SplineInterpolator, self).setup()
+        points = np.asarray(self.get_points())
+
+        if len(points) < (self.degree + 1):
+            raise ValueError(f"Minimum {self.degree + 1} "
+                             f"keyframes must be set in order to use "
+                             f"{self.degree}-degree spline")
+
+        self.tck = interpolate.splprep(points.T, k=self.degree, full_output=1,
+                                       s=self.smoothness)[0][0]
+        self.linear_lengths = []
+        for x, y in zip(points, points[1:]):
+            self.linear_lengths.append(
+                math.sqrt((x[1] - y[1]) * (x[1] - y[1]) +
+                          (x[0] - y[0]) * (x[0] - y[0])))
+
+    def get_points(self):
```

Comment:
Why this function is only for this interpolator?

Reply:
I needed the points to calculate the linear lengths only for spline interpolator.
Yeah, it was being used once, and it was simple, so I just added the implementation of it to the setup method directly.